euphonic/qpoint_frequencies.py

```diff
@@ -93,28 +94,45 @@ def calculate_dos(self, dos_bins: Quantity) -> Spectrum1D:
         dos_bins
             Shape (n_e_bins + 1,) float Quantity. The energy bin edges
             to use for calculating the DOS
+        mode_widths
+            Shape (n_qpts, n_branches) float Quantity in energy units.
+            The broadening width for each mode at each q-point, for
+            adaptive broadening
 
         Returns
         -------
         dos
             A spectrum containing the energy bins on the x-axis and dos
             on the y-axis
         """
-
         freqs = self._frequencies
-        dos_bins_unit = dos_bins.units
+        n_modes = self.frequencies.shape[1]
         # dos_bins commonly contains a 0 bin, and converting from 0 1/cm
         # to 0 hartree causes a RuntimeWarning, so suppress it
         with warnings.catch_warnings():
             warnings.filterwarnings('ignore', category=RuntimeWarning)
-            dos_bins = dos_bins.to('hartree').magnitude
-        weights = np.repeat(self.weights[:, np.newaxis],
-                            self.frequencies.shape[1],
-                            axis=1) / np.sum(self.weights)
-        dos, _ = np.histogram(freqs, dos_bins, weights=weights)
+            dos_bins_calc = dos_bins.to('hartree').magnitude
+        if mode_widths is not None:
+            from scipy.stats import norm
+            dos_bins_calc = Spectrum1D._bin_edges_to_centres(dos_bins_calc)
+            dos = np.zeros(len(dos_bins_calc))
+            mode_widths = mode_widths.to('hartree').magnitude
+            # Set limit on very small widths of, say, 5e-7 hartree
+            # (~0.013meV) due to divergent behaviour
+            mode_widths = np.maximum(mode_widths, 5e-7)
```

Comment:
I'm not sure, there's a few places in Euphonic where there are magic numbers like this and I don't really know what to do with them... Do we want to be able to override them? I feel like don't really want to add an argument because 99% of the time you won't want to change it, but this specific case isn't used anywhere else so it wouldn't necessarily make sense to make it more global...

Reply:
In this case, it could be an optional argument with a default value.

One approach would be what Abins does which is to define constants in one module and import them elsewhere. But that's more suited to constants that are used in multiple places. I'd prefer the kwarg approach in this particular case.

Reply:
Kwarg added